src/main/java/hudson/plugins/ansicolor/ColorConsoleAnnotator.java

```diff
@@ -0,0 +1,135 @@
+    }
+
+    @Override
+    public ConsoleAnnotator<Object> annotate(Object context, MarkupText text) {
+        String s = text.getText();
+        if (s.indexOf('\u001B') != -1) {
+            AnsiColorMap colorMap = Jenkins.get().getDescriptorByType(AnsiColorBuildWrapper.DescriptorImpl.class).getColorMap(colorMapName);
+            CountingOutputStream outgoing = new CountingOutputStream(new NullOutputStream());
+            class EmitterImpl implements AnsiAttributeElement.Emitter {
+                CountingOutputStream incoming;
+                int adjustment;
+                int lastPoint = -1; // multiple HTML tags may be emitted for one control sequence
+                @Override
+                public void emitHtml(String html) {
+                    LOGGER.finest("emitting " + html + " @" + incoming.getCount());
+                    text.addMarkup(incoming.getCount(), html);
+                    if (incoming.getCount() != lastPoint) {
+                        lastPoint = incoming.getCount();
+                        int hide = incoming.getCount() - outgoing.getCount() - adjustment;
+                        LOGGER.finest("hiding " + hide + " @" + (outgoing.getCount() + adjustment));
+                        text.addMarkup(outgoing.getCount() + adjustment, outgoing.getCount() + adjustment + hide, "<span style=\"display: none\">", "</span>");
+                        adjustment += hide;
+                    }
+                }
+            }
+            EmitterImpl emitter = new EmitterImpl();
+            CountingOutputStream incoming = new CountingOutputStream(new AnsiHtmlOutputStream(outgoing, colorMap, emitter));
+            emitter.incoming = incoming;
+            try {
+                byte[] data = s.getBytes(charset);
+                for (int i = 0; i < data.length; i++) {
+                    // Do not use write(byte[]) as offsets in incoming would not be accurate.
+                    incoming.write(data[i]);
+                }
+            } catch (IOException x) {
+                LOGGER.log(Level.WARNING, null, x);
+            }
+            LOGGER.finer(() -> "\"" + StringEscapeUtils.escapeJava(s) + "\" → \"" + StringEscapeUtils.escapeJava(text.toString(true)) + "\"");
+        }
+        return this;
+    }
+
+    @Extension
+    public static final class Factory extends ConsoleAnnotatorFactory<Object> {
+
+        @Override
+        public ConsoleAnnotator<Object> newInstance(Object context) {
+            LOGGER.fine("context=" + context);
+            if (context instanceof Run) {
+                ColorizedAction action = ((Run) context).getAction(ColorizedAction.class);
+                if (action != null) {
+                    return new ColorConsoleAnnotator(action.colorMapName, ((Run) context).getCharset().name());
+                }
+            } else if (Jenkins.get().getPlugin("workflow-api") != null && context instanceof FlowNode) {
```

Comment:
Did not try to verify behavior when the plugin is not in fact installed. Generally speaking, `optional` deps are dangerous. I would rather just make it a hard dep.

Reply:
I did some testing with the plugins not installed and things seem to work fine. New freestyle builds are still colored correctly, and executing `Jenkins.get().getPlugin("workflow-api")` in the script console returns null, so the short-circuiting behavior here looks fine to me.